src/logger/lwt_log_core.ml

```diff
@@ -101,31 +112,30 @@ let split pattern =
   in
   loop 0
 
-
 let rules = ref []
 
-let load_rules' str =
+let load_rules' str fail_on_error =
   let rec loop = function
-    | [] ->
-      []
-    | (pattern, level) :: rest ->
-      let pattern = split pattern in
-      match (String.lowercase [@ocaml.warning "-3"]) level with
-        | "debug" -> (pattern, Debug) :: loop rest
-        | "info" -> (pattern, Info) :: loop rest
-        | "notice" -> (pattern, Notice) :: loop rest
-        | "warning" -> (pattern, Warning) :: loop rest
-        | "error" -> (pattern, Error) :: loop rest
-        | "fatal" -> (pattern, Fatal) :: loop rest
-        | level -> log_intern "invalid log level (%s)" level; loop rest
+  | [] -> []
+  | (pattern, level_str) :: rest ->
+    let pattern = split pattern in
+    let level = level_of_string level_str in
+    match level with
+    | Ok level -> (pattern, level) :: loop rest
+    | Pervasives.Error msg ->
+      if fail_on_error then raise (Failure msg)
+      else log_intern "invalid log level (%s)" level_str; loop rest
   in
   match Lwt_log_rules.rules (Lexing.from_string str) with
-    | None -> Printf.eprintf "Invalid contents of the LWT_LOG variable\n%!"
-    | Some l -> rules := loop l
+  | None ->
+    if fail_on_error then raise (Failure "Invalid log rules")
+    else Printf.eprintf "Invalid contents of the LWT_LOG variable\n%!"
```

Comment:
Since we're fixing this up anyway, I think the reference to `LWT_LOG` should be removed – this code can be called when not reading `LWT_LOG`. Alternatively, one could pass some kind of `from_lwt_log` parameter to control what this message looks like... but it may be kind of unnecessary. Ideally we would push the generation of the message to the caller, but that would require a breaking change :/